kernel/kernel-api/src/main/java/io/delta/kernel/Transaction.java

```diff
@@ -0,0 +1,207 @@
+     *
+     * @param tableClient    {@link TableClient} instance.
+     * @param stagedData     Iterator of data actions to commit. These data actions are generated by
+     *                       the `Transaction.stageAppendOnlyData` or manually created. Manual
+     *                       creation of data actions is useful when has the data files and want to
+     *                       directly inject them into the table. One important note is that the
+     *                       connector should be able to provide the same content if the transaction
+     *                       commit fails and needs to be retried.
+     * @param maximumRetries If the transaction is retried, return the previous status of the
+     *                       transaction. This helps the Kernel reuse the conflict resolution work
+     *                       done in the previous attempt.
+     * @return {@link TransactionCommitStatus} status of the successful transaction.
+     * @throws ConcurrentWriteException when the transaction has encountered a non-retryable
+     *                                  conflicts or exceeded the maximum number of retries reached.
+     *                                  The connector needs to rerun the query on top of the latest
+     *                                  table state and retry the transaction.
+     */
+    TransactionCommitStatus commitWithRetries(
+            TableClient tableClient,
+            CloseableIterable<Row> stagedData,
+            int maximumRetries
+    ) throws ConcurrentWriteException;
+
+    /**
+     * Given the logical data that needs to be written to the table, convert it into the required
+     * physical data depending upon the table Delta protocol and features enabled on the table.
+     * Kernel takes care of adding any additional column or removing existing columns that doesn't
+     * need to be in physical data files. All these transformations are driven by the Delta protocol
+     * and table features enabled on the table.
+     * <p>
+     * The given data should belong to exactly one partition. It is the job the connector to do
+     * partitioning of the data before calling the API. Partition values are provided as map of
+     * column name to partition value (as {@link Literal}). If the table is an un-partitioned table,
+     * then map should be empty.
+     *
+     * @param tableClient      {@link TableClient} instance to use.
+     * @param transactionState The transaction state
+     * @param dataIter         Iterator of logical data to transform to physical data. All the data
+     *                         in this iterator should belong to one physical partition.
+     * @param partitionValues  The partition values for the data. If the table is un-partitioned,
+     *                         the map should be empty
+     * @return Iterator of physical data to write to the data files.
+     */
+    static CloseableIterator<FilteredColumnarBatch> transformLogicalData(
```

Comment:
can we make it clearer that this is for a single partition? you can push back and hold that the API docs are sufficient

`transformLogicalDataForSinglePartition` ?

`transformLogicalDataUnpartitioned`?